SW4STM32/i2cio/Application/User/encoder.c

```diff
@@ -61,21 +61,19 @@ static inline bool detect(bool x1, bool x2, bool y1, bool y2)
 	return (x2 ^ y1) & ~(x1 ^ y2);
 }
 
-void encoderCapture(){
+void encoderCapture() {
 	for (uint8_t i = 0; i < MAX_ENCODER_COUNT; ++i){
 		if (encoders[i].enabled){
 			bool stateA = pinState(encoders[i].pinA);
 			bool stateB = pinState(encoders[i].pinB);
-			bool _stateA = encoders[i].lastStateA;
-			bool _stateB = encoders[i].lastStateB;
-			
-			if (detect(_stateA, _stateB, stateA, stateB))
-				encoders[i].unreadedValue++;
-			else if (detect(_stateB, _stateA, stateB, stateA))
-				encoders[i].unreadedValue--;
-
-			encoders[i].lastStateA = stateA;
-			encoders[i].lastStateB = stateB;
+			if(((encoders[i].stateA & 1) != (int)stateA) || ((encoders[i].stateB & 1) != (int)stateB)) { // если состояние изменилось
```

Comment:
Too complex condition. Extract to variables, use early exit:

```cpp
uint8_t oldStateA = encoders[i].stateA;
uint8_t oldStateB = encoders[i].stateB;
bool changedA = oldStateA & 0x1 != (uint8_t)stateA;
bool changedB = oldStateB & 0x1 != (uint8_t)stateB;
if (!changedA && !changedB)
    continue;
```

Reply:
Ok. I use #defines for it.